### PR TITLE
Make msghdr.msg_name optional

### PR DESCRIPTION
**Pull request by redbaron**
sendmsg can be used on connected sockets without msg_name.

### CODE REVIEW

**Comment by FrankReh** [comment]
These changes may be a step in the right direction but they are not complete. The branch doesn't compile.

Please run `cargo build` and `cargo test` on your branch before starting a PR. Also `cargo fmt` and `cargo clippy`. All with the most recent compiler. In this case, you can make additional commits to your branch. See all those commands pass and then push up to GitHub again.

Here, even the file src/io/sendmsg_zc.rs which you changed won't compile, let along the entire library not compiling.

And while your comment about sendmsg is correct, please include a link to an online manpage from some place like arch linux so anyone reading through this can be starting with the same understanding.

Thank you.

**Comment by redbaron** [comment]
I am really sorry for incomplete PR, IDK how it happened. Pushed fixed version.

File: src/net/udp.rs
Comment (redbaron) [inline src/net/udp.rs:246]: Regarding comment above:

> /// Note: Using fixed buffers [#54](https://github.com/tokio-rs/tokio-uring/pull/54), avoids the page-pinning overhead

I don't think it is true. If I read Linux kernel code correctly, fixed buffers can be used with send_zc only:
https://elixir.bootlin.com/linux/v6.1.10/source/io_uring/net.c#L983

it makes sense, SQE has only one buf_index, but msghdr can provide multiple buffers in msg_iov

Comment (FrankReh) [reply src/net/udp.rs:246]: You are right. Good catch. Please remove that comment line. I'll thoroughly review tomorrow morning.

By the way, is there an actual unit test that had used an address that wasn't necessary that you could copy and show that the None option works as intended?

Comment (redbaron) [reply src/net/udp.rs:256]: > By the way, is there an actual unit test that had used an address that wasn't necessary that you could copy and show that the None option works as intended? 

I updated one of the doctests, let me know if you would like to move it elsewhere

Comment (redbaron) [reply src/net/udp.rs:256]: > In fact, I don't see this crate supporting fixed buffers for send_zc

Yeah, there should be `send_fixed_zc` with  `BoundedBuf<Buf = FixedBuf>` and corresponding op, which prepares  IORUNG_SEND_ZC SQE with necessary flags set.

Comment (FrankReh) [reply src/net/udp.rs:256]: Would you mind opening two issues for the missing send_zc_fixed in io_uring and tokio_uring. I think we should follow the naming convention used in liburing when possible. In this case, they had chosen to add _fixed after the send_zc.

Doesn't mean you are on the hook for submitting PRs for them.

File: src/net/udp.rs
Comment (FrankReh) [inline src/net/udp.rs:211]: Rather than modify this function to take an Optional address, please revert the changes to this comment and the argument and create a new function right above it, `send`. This will more closely mirror the Linux man pages for send and sendto.

The Optional address for the sendmsg_zc is good the way you've changed it in this PR.

This PR also highlights that we don't have a regular `sendmsg` for UDP. You don't have to do anything about that but you or I should open an Issue about that.

File: src/net/udp.rs
Comment (FrankReh) [inline src/net/udp.rs:259]: Please update the sendmsg_zc comment to say that None can be used when the socket has been connected and that it may be more efficient when sending multiple datagrams to the same address. I don't think you have to mention why it may be more efficient. I figure it's enough to trace through the code and see that a boxed value can be skipped for the uring operation and I haven't looked, but the kernel may also have already done its address parsing when the socket is connected so it could help on the kernel side too.

**Review by FrankReh — CHANGES_REQUESTED** [review]
Looking good. Just asking for some small changes, all in net/udp.rs I think.

File: src/net/udp.rs
Comment (redbaron) [reply src/net/udp.rs:211]: You want `send` to not accept address and `send_to` to always require address, right?

Comment (FrankReh) [reply src/net/udp.rs:211]: Yes. That would mirror what I see from the Linux man page for send and sendto. Does that sound reasonable to you? I haven't missed either but I haven't had to send flags along with any send yet so I'm just looking for what makes the most sense at least in some vein. With all these ways of sending data, there's bound to be opinions.

Comment (FrankReh) [reply src/net/udp.rs:211]: Only make these changes to net/udp.rs. Both the send and send_to in that file can then use the common io::send_to as you've changed it to optionally take an address.

File: src/net/udp.rs
Comment (redbaron) [reply src/net/udp.rs:259]: Maybe create new struct UdpSocketConnected to be returned from UdpSocket::connect? It will have mostly same methods, but `sendmsg_zc` and `send_to` will be not taking `socket_addr` at all?

Comment (FrankReh) [reply src/net/udp.rs:259]: Please add that as a developer's comment so the idea isn't lost but I would rather not try to enforce this with a compiler. The io_uring device should return an error in those cases.

File: src/net/udp.rs
Comment (redbaron) [reply src/net/udp.rs:211]: It wont be different from `UdpSocket::write`, do we need it at all then? Maybe leave non ZC methods as they are?

File: src/net/udp.rs
Comment (redbaron) [reply src/net/udp.rs:259]: Compile time error is better than runtime. Isn't it better not to provide incorrect API at all, rather than documenting quirks and asking users to be careful?

File: src/net/udp.rs
Comment (FrankReh) [reply src/net/udp.rs:211]: I think we should support send_to and send and write for udp. Send and write use different uring opcodes and over time, we will have a builder for some or all of these operations to let the user get at more of the optional fields defined by the io-uring's opcodes and this is a good time to introduce send since it takes no address while send_to does. Even before we expose the options through a builder, the io_uring device offers different opcodes and we can let the user of this library decide which they prefer.

File: src/net/udp.rs
Comment (redbaron) [reply src/net/udp.rs:264]: https://github.com/tokio-rs/io-uring/issues/185
https://github.com/tokio-rs/tokio-uring/issues/235

File: src/net/udp.rs
Comment (redbaron) [reply src/net/udp.rs:221]: added `UdpSocket::send<T: BoundedBuf>(&self, buf: T) -> crate::BufResult<usize, T>`

File: src/net/udp.rs
Comment (redbaron) [reply src/net/udp.rs:270]: Added comment as  requested

File: src/net/udp.rs
Comment (FrankReh) [inline src/net/udp.rs:221]: Since your comment here says `or to a previously connected remote peer`, want to add that `send` might be preferred in the case of a previously connected remote peer as it saves a box?

**Review by FrankReh — CHANGES_REQUESTED** [review]

File: src/net/udp.rs
Comment (redbaron) [reply src/net/udp.rs]: I removed comment about connected sockets, because `UdpSocket::send_to`can't be used without SocketAdddr now

**Review by ollie-etl — APPROVED** [review]

**Review by FrankReh — APPROVED** [review]

**Comment by FrankReh** [comment]
Thank you!

**Comment by redbaron** [comment]
> Thank you!

I didn't realise you'd merge it straight away, my intention was to get approval then squash all changes. sorry for many useless commits

**Comment by FrankReh** [comment]
We squash the changes for every PR in this repo. At least we mean to. There may have been some that slipped through. All the commits weren't a problem. The GitHub merge will do the squash for us if we like. I did end up pulling this branch locally to be able to look through things when there were problems. The final merge from master I didn't see needing to wait for you as it might have taken another day for the turn around. And there are a number of outstanding PRs that can get merged when another maintainer has a look so I wanted as much committed and remerged to the other PRs as quickly as possible.

**Pull request merged**
